### ISSUE TITLE
[Feature Request] second level sub-menu for menu item

### ISSUE THREAD

**Issue by natural-harmonia-gropius** [tomasklaen/uosc#475]
To make "pin" or "delete" a menu item works more naturally, allow me to address https://github.com/natural-harmonia-gropius/recent-menu/issues/9#issuecomment-1467323505

![屏幕截图 2023-03-16 002612](https://user-images.githubusercontent.com/50797982/225376658-fd232b24-1bb9-4a43-8a19-a74351a3a871.png)

Not in English 😓.

Anyway, it's a second level sub-menu for menu item by right click. I want this in uosc, add `Menu.items_second` or `open-submenu <json>` will be good.

**Comment by hooke007** [comment]
Just a idea, you could consider that adding buttons at the end the playlist items ? (which could bind the key_binding we already had https://github.com/tomasklaen/uosc/commit/f039af57324a74b6f92eac894b02a2226312e9cd.)

Just look like this:
<img width="465" alt="Snipaste_2023-03-29_00-28-14" src="https://user-images.githubusercontent.com/41094733/228389476-02c108af-3273-4fca-bc2e-ad9101cbecb4.png">

**Comment by tomasklaen** [comment]
Yeah, one of the drivers for #483 is so that I can add buttons to menu items.

**Comment by christoph-heinrich** [comment]
> something like this could be achieved by exposing key meta for menu implementers.

The most backwards compatible way of doing this might be by changing the type of an item value.

```
value: string | string[] | { main?: string | string[] ; ctrl?: string | string[] ; shift?: string | string[] ; alt?: string | string[] }
```
If we also want to support all possible permutations of modifier + selection that would be 8 possible fields.

Alternatively we could make a new `on_selection` alternative to `value` that sends a command and appends a list of pressed modifiers, but then we'd always have to listen to all possible selection combinations, where as in the first variant we can listen to only the required ones.

memo would benefit from this as it would allow appending to the playlist instead of opening the file.

**Comment by tomasklaen** [comment]
My current plan is to have a separate event emitter that when registered, takes over just triggering a value as a command. Something like defining a script listener for activation events for current menu:

```lua
{type = 'my-menu-type', listener = 'my-script-name', items = {...}}
```

When defined this would then send not just activation events, but any navigation and modifier related ones to the listener as json with a message like:

```
script-message-to my-script-name menu-press my-menu-type {data_json}
```

Where data would have all of this info:

```lua
{
  value = 'menu item value',
  key = 'enter', -- What key was pressed, left mouse button normalizes to enter
  action = 'main', -- `name` field of a button that was pressed, 'main' reserved for the main one
  modifier_id = 'alt+shift', -- Alphabetically ordered list of pressed modifiers separated by plus sign
  key_id = 'alt+shift+enter', -- Same as ^ plus the key
  alt = true,
  ctrl = false,
  shift = true,
}
```

**Comment by christoph-heinrich** [comment]
> ```lua
>   modifier_id = 'alt+shift', -- Alphabetically ordered list of pressed modifiers separated by plus sign
>   key_id = 'alt+shift+enter', -- Same as ^ plus the key
>   alt = true,
>   ctrl = false,
>   shift = true,
> ```

Why the same thing multiple times?

**Comment by tomasklaen** [comment]
I'm coming from a decade of implementing interactions in various environments, and some times I need key id, other times I need modifier flags. You just don't know how and what will people want to implement. Maybe they'll just want to listen to a specific key and compound actions based on pressed modifiers. But yeah, most people will only be interested in `key_id`.

**Comment by christoph-heinrich** [comment]
I don't see how that's more flexible then
```
  key = 'enter',
  alt = true,
  ctrl = false,
  shift = true,
```
but ok, I guess it doesn't hurt to provide more stuff.

**Comment by tomasklaen** [comment]
If you only have flag booleans, you need to add all of them to your condition, or you'll trigger on shortcuts you don't want. Example:

```lua
-- You want to do something on `ctrl+up`, so you do a naive:
if meta.key == 'up' and meta.ctrl then
  -- this also fires for `ctrl+shift+up`, `alt+ctrl+up`, etc.
end

-- So you need to do this
if meta.key == 'up' and meta.ctrl and not meta.shift and not meta.alt and not meta.cmd then
  -- which is annoying
end

-- With `key_id` you only need this simple check:
if meta.key_id == 'ctrl+up' then end
```

**Comment by christoph-heinrich** [comment]
Fair enough.

**Comment by tomasklaen** [comment]
So menus for menu items is not gonna happen, but with the new menu callback mode API you can add item actions, and branch even further with modifiers. Hope that helps.
